Refactor: Add WordPressMedia

kean opened the pull request:
This PR enables a bit of modularization by extracting some obvious stuff: `ImageDownloader`. The framework name is `WordPressMedia` because the idea is to put everything media-related there, including exporters in the future. 

The main advantage of this change is that you can now compile and run tests for this new target nearly instantly, so it will encourage us to add more tests and make it easier to maintain existing ones.

Changes:
- Add `WordPressMedia` target
- Move `ImageDownloader` and the types it depend on to the new target
- Update `ImageDownloader` tests to use the new Swift Testing framework
- Add `WordPressTesting` target to share test code and resources across test targets

## Regression Notes
1. Potential unintended areas of impact


2. What I did to test those areas of impact (or what existing automated tests I relied on)


3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dangermattic commented:
<!--
  0 Errors
  4 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          4 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

crazytonyli commented:
Do you think it's worthwhile enabling strict concurrency checks on the new targets? If so, we can wait for https://github.com/wordpress-mobile/WordPress-iOS/pull/23838 to be merged.

crazytonyli commented on Modules/Sources/WordPressTesting/Empty.swift:
```diff
@@ -0,0 +1,5 @@
+import Foundation
+
+extension Bundle {
+    public static var test: Bundle { Bundle.module }
```
`test` seems too generic. What do you think of naming it `wordPressTesting`?

kean replied:
I didn't rename it, so I'd suggest sticking with this name for now. I think it's OK because you are only ever going to import it in a subset of files in tests targets. It doesn't affect anything else.

<img width="1631" alt="Screenshot 2024-12-12 at 6 03 13 PM" src="https://github.com/user-attachments/assets/234abe6a-dfc3-4461-a540-0fe3e82869f1" />

kean commented:
> Do you think it's worthwhile enabling strict concurrency checks on the new targets?

Sure. New modules like this should start with Swift 6 enabled. I approved the PR you referenced.

crazytonyli commented:
Thanks! The PR has been merged.

kean committed:
Add WordPressMedia

kean committed:
Add Swift 6 support

kean commented:
I rebased it (with no changes) and added Swift 6 support in https://github.com/wordpress-mobile/WordPress-iOS/pull/23890/commits/ea242025eb1f69fe52f75de2db1603e54f8a4a2a. The changes should be pretty straightforward. I added `@ImageDownloaderActor` to synchronize more types on it in the future (prefetcher). `ImageDownloader (Closures)` are no longer in the module.

kean commented:
I'm going to merge this later _after_ 24.6 is properly cut-off.

kean committed:
Fix build
